SOC/models/ofc.py

```diff
@@ -0,0 +1,100 @@
+"""Implements the OFC model."""
+
+from SOC import common
+import numpy as np
+import numba
+import random
+
+class OFC(common.Simulation):
+    """Implements the OFC model."""
+    
+    def __init__(self, L: int, critical_value: float = 1., conservation_lvl: float = 0.25):
+        """
+        :param L: linear size of lattice, without boundary layers
+        :type L: int
+        :param critical_value: 1.0 by default - above this value, nodes start toppling
+        :type critical_value: float
+        :param conservation_lvl: 0.25 by default - fraction of the force from a toppling site going to its neighbour   # 0.25 -> full force distributed (if 4 neighbours)
+        :type conservation_lvl: float
+        """
+        super().__init__(L)
+        self.critical_value = critical_value
+        self.values = np.random.rand(self.L_with_boundary, self.L_with_boundary) * self.critical_value
+        self.conservation_lvl = conservation_lvl
+        
+    def drive(self):
+        """
+        Drive the simulation by adding force from the outside.
+
+        """
+        #increasing all values by the amount needed by the site with highest value to reach the critical_value
+        location_of_max = np.unravel_index(np.argmax(self.values, axis=None), self.values.shape)
+        self.values += self.critical_value - self.values[location_of_max]
```

Comment:
Czyli, jeśli dobrze rozumiem:

1. Znajdujesz maksymalną wartość w `self.values` (co, swoją drogą, chyba można zrobić po prostu przez `np.max(self.values)`)
2. Dodajesz do WSZYSTKICH wejść w `self.values` różnicę między wartością krytyczną a maksimum z tej tablicy

Tylko że to będzie chyba wolne, bo jak mamy tablicę N na N wejść, to musisz zrobić N^2 operacji na tej tablicy. Z tego co pamiętam z Dobrej Referencji, można to też robić przez opuszczanie w dół wartości krytycznej - to jedna operacja zamiast N^2!

Reply:
Rzeczywiście tak będzie efektywniej. Chwilowo jednak, jeszcze w oryginalnym podejciu, mam problem właśnie z active_sites, które od którejś iteracji drivingu pozostają całe False.